rfcs/2020-11-02-remap-strict-return-value-types.md

```diff
@@ -0,0 +1,278 @@
+
+1. The literal expression `"bar"`, which resolves to a `Value::String`.
+2. The assignment expression `.foo = …`, which resolves to the value of the
+   expression on the right side of the assignment operator.
+
+This example program, when executed, always resolves to a `String` value.
+
+Another example program is `floor(.bar)`. This program again consists of two
+expressions:
+
+1. The query path expression `.bar`, which resolves to any value the `bar`
+   object field contains.
+2. The function expression `floor`, which resolves to either an `Integer`, or
+   `Float`.
+
+Since the `floor` function is the final expression that runs in the program, the
+outcome of a _successful execution_ of the program is either an `Integer` or
+`Float`.
+
+**expression trait**
+
+Expressions are implemented using the `Expression` trait:
+
+```rust
+pub trait Expression: Send + Sync {
+    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>>;
+}
+```
+
+Expressions can be fallible (not covered in this RFC), can return `None` (not
+covered in-depth in this RFC), or return a resolved `Value`.
+
+**defining the expected resolved values**
+
+To allow the execution context of a program to determine the outcome of running
+the program, we extend the `Expression` trait as follows:
+
+```rust
+pub trait Expression: Send + Sync {
+    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>>;
+
+    /// Describes which [`ValueKind`]s the expression can resolve to.
+    ///
+    /// bike-shedding of function and enum names welcome.
+    fn resolves_to(&self) -> ResolveKind;
+}
+```
+
+We define `ResolveKind` as follows:
+
+```rust
+pub enum ResolveKind {
+    /// The expression can resolve to any value at runtime.
+    Any,
+
+    /// The expressions resolves to one of the defined [`ValueKind`]s.
+    OneOf(Vec<ValueKind>),
+
+    /// If the expression succeeds, it might resolve to a value, but doesn't
+    /// have to.
+    Maybe(Box<ResolveKind>),
+}
+```
+
+The `ResolveKind` enum allows us to determine if an expression resolves to a
+predefined list of values. It also defines if the expression can resolve to
+"nothing" (e.g. `None`).
+
+**transient expressions**
+
+Some expressions don't have their own return value, but instead resolve to the
+return value of an expression they themselves execute.
+
+One such an example is the `IfStatement` expression, which resolves to either
+the `true_expression` or the `false_expression`.
+
+Because expressions are initialized at compile-time, we are able to have such
+"transient expressions" return the values they resolve to based on whatever
+expressions they themselves hold:
+
+```rust
+impl Expression for IfStatement {
+    fn resolves_to(&self) -> ResolveKind {
+        let true_resolves = self.false_expression.resolves_to();
+        let false_resolves = self.false_expression.resolves_to();
+
+        // return the combined set of true and false resolves.
```

Comment:
"combined set" here means that if either one returns `Any` or `Maybe`, then this expression will also return any or maybe. Other than that, the two arrays of value kinds are merged together.